util/fpga/splice.py

```diff
@@ -0,0 +1,385 @@
+                tmpbit = f'{tmp}/{i}.bit'
+                self.updatemem(infile, mmi, f'{tmp}/{i}.mem', tmpbit)
+                infile = tmpbit
+
+            # Update the usr_access_value to make this bitstream uniquely identifiable.
+            self.update_usr_access(infile, f'{tmp}/updated.bit')
+
+            # Copy the final bitstream into the cache and return the bitstream content.
+            with open(f'{tmp}/updated.bit', 'rb') as f:
+                content = f.read()
+                if outfile:
+                    outfile.seek(0)
+                    outfile.truncate()
+                    outfile.write(content)
+                    outfile.seek(0)
+                return content
+
+
+def main(args):
+    mmi = args.mmi.split(',')
+    data = args.data.split(',')
+    if len(data) == 0:
+        logging.error('No data files to splice')
+        return 1
+    if len(data) != len(mmi):
+        logging.error(
+            'There must be a 1:1 correspondence between data and mmi files')
+        return 1
+
+    cachedir = os.path.expanduser(args.cache)
+    splicedir = os.path.join(cachedir, 'splice')
+    os.makedirs(splicedir, exist_ok=True)
+
+    tools = Tools(args)
+    key = tools.hash([args.bitstream] + data + mmi)
+    logging.info(f'Checking cache for bitstream {key}')
+    cache = CacheControlFile(f'{cachedir}/splices.json')
+
+    # Expire old bitstreams to prevent the cache from getting too big.
+    expired = cache.expire(args.cache_entries)
+    tools.remove_files(
+        os.path.join(cachedir, ex['filename']) for ex in expired.values())
+
+    splice = cache.get(key)
+    if splice is None:
+        logging.info(f'Bitstream doesn\'t exist.  Creating it')
+        splice = cache.create(key, f'splice/{key}.bit')
+    else:
+        logging.info(f'Bitstream is at {splice["filename"]}')
+
+    # We open and lock the cache's bitstream file so that we can close
+    # the cache and unlock other instances while we generate a bitstream.
+    # - If the other instances are generating different bitstreams, they
+    #   can proceed in parallel.
+    # - If the other instances want to generate the same bitstream, they
+    #   will block on the lock and used the bitstream created by the instance
+    #   holding the lock.
+    filename = os.path.join(cachedir, splice['filename'])
+    bitstream = open_locked(filename, binary=True)
+    content = bitstream.read()
+    digest = hashlib.sha256(content).hexdigest()
+
+    if digest != splice['digest']:
+        # If the digests don't match, we try once to refresh the cache.
+        # This handles the case when a second process starts while the first
+        # process is still splicing and the second process gets an empty
+        # digest.  When the first process finishes, it will update the cache
+        # and unlock the bitstream file.  If the second process refreshes
+        # here, it will then see the newly computed digest.
+        cache.refresh()
```

Comment:
at this point, we have locked both the:
- cache control file, and
- the bitstream file

If another process is working on generating the same bitstream that this process would've produced, i.e. it was somewhere between lines 359 to 368 (inclusive), wouldn't there be a race condition here? Since `cache.refresh()` releases and reacquires the lock on the cache control file, you would be hoping that the other process is able to reacquire the cache control lock first, no?

Reply:
I've restructured the ordering of the locking as in our offline discussion: lock the tentative bitstream file first, then the cache control file.